modules/i2c/slavex.inc.c

```diff
@@ -6,7 +6,203 @@
  * The X_(p,s) macro must be defined before including.
  * It is automatically undefined at the end of this file.
  */
+#include <avr/interrupt.h>
+#include <stddef.h>
 
-#error Slave mode not supported yet
+#define I2CX(s) X_(I2C,s)
+#define i2cX(s) X_(i2c,s)
+#define TWIX X_(TWI,)
+#define twiX(s) X_(TWI,s)
 
+// i2cX singleton
+i2cs_t i2cX();
+
+void i2cX(_init)(void) {
+  i2cX().process_data = NULL;
+  i2cX().bytes_received = 0;
+  i2cX().bytes_transmit = 0;
+  i2cX().status = I2CS_STATUS_READY;
+  i2cX().result = I2CS_RESULT_UNKNOWN;
+
+  TWIX.SLAVE.CTRLA = (I2C_INTLVL<<TWI_SLAVE_INTLVL_gp) |
+                     TWI_SLAVE_DIEN_bm |
+                     TWI_SLAVE_APIEN_bm |
+                     TWI_SLAVE_ENABLE_bm;
+  TWIX.SLAVE.ADDR = (I2CX(_ADDRESS)<<1);
+}
+
+
+// Register callback
+void i2cX(s_register_callback)( void (*process_data_function)(void) ) {
+  i2cX().process_data = process_data_function;
+}
+
+/* @brief I2C transaction finished function.
+ *  Prepares module for new transaction.
+ *  @bparam result The result of the transaction.
+ */
+static void i2cX(_transaction_finished(uint8_t result)) {
+  i2cX().result = result;
+  i2cX().status = I2CS_STATUS_READY;
+}
+
+/* @brief i2c slave transmit interrupt handler.
+ *  Handles i2c slave transmit transactions and responses.
+ */
+static void i2cX(_transmit_handler(void)) {
+  /* If NACK, slave transmit transaction finished. */
+  if ((i2cX().bytes_transmit > 0) && (TWIX.SLAVE.STATUS &
+                                 TWI_SLAVE_RXACK_bm)) {
+
+    TWIX.SLAVE.CTRLB = TWI_SLAVE_CMD_COMPTRANS_gc;
+    i2cX(_transaction_finished(I2CS_RESULT_OK));
+  }
+  /* If ACK, master expects more data. */
+  else {
+    if (i2cX().bytes_transmit < I2CS_SEND_BUFFER_SIZE) {
+      uint8_t data = i2cX().transmit_data[i2cX().bytes_transmit];
+      TWIX.SLAVE.DATA = data;
+      i2cX().bytes_transmit++;
+      i2cX().result = I2CS_RESULT_TRANSMIT;
+
+      /* Send data, wait for data interrupt. */
+      TWIX.SLAVE.CTRLB = TWI_SLAVE_CMD_RESPONSE_gc;
+    }
+    /* If buffer overflow. */
+    else {
+      TWIX.SLAVE.CTRLB = TWI_SLAVE_CMD_COMPTRANS_gc;
+      i2cX(_transaction_finished(I2CS_RESULT_BUFFER_OVERFLOW));
```

Comment:
Should be `i2cX(_transaction_finished)(I2CS_RESULT_BUFFER_OVERFLOW);`

Similar occurrences below.